src/protocols/bdx/AsyncTransferFacilitator.cpp

```diff
@@ -0,0 +1,208 @@
+ * message over the exchange context, otherwise it calls the HandleTransferSessionOutput method implemented by the subclass to
+ * handle the BDX message.
+ */
+void AsyncTransferFacilitator::ProcessOutputEvents()
+{
+    if (mProcessingOutputEvents)
+    {
+        ChipLogDetail(BDX, "ProcessOutputEvents: Still getting and processing output events from a previous call. Return.");
+        return;
+    }
+
+    mProcessingOutputEvents = true;
+
+    // Get the next output event and handle it based on the type of event.
+    // If its of type kMsgToSend send it over the exchange, otherwise call the HandleTransferSessionOutput
+    // virtual method that must be implemeted by the subclass of this class to handle the BDX message.
+    TransferSession::OutputEvent outEvent;
+
+    mTransfer.GetNextAction(outEvent);
+    while (outEvent.EventType != TransferSession::OutputEventType::kNone)
+    {
+        if (outEvent.EventType == TransferSession::OutputEventType::kMsgToSend)
+        {
+            CHIP_ERROR err = SendMessage(outEvent.msgTypeData, outEvent.MsgData);
+
+            // If we failed to send the message across the exchange, there is no way to let the other side know there was an
+            // error so call DestroySelf so the exchange can be released and the other side will be notified the exchange is
+            // closing and will clean up.
+            if (err != CHIP_NO_ERROR)
+            {
+                DestroySelf();
+                return;
+            }
+
+            // If we send out a status report across the exchange, schedule a call to DestroySelf() at a little bit later time
+            // since we want the message to be sent before we clean up.
+            if (outEvent.msgTypeData.HasMessageType(Protocols::SecureChannel::MsgType::StatusReport))
+            {
+                DestroySelf();
+                return;
+            }
+        }
+        else
+        {
+            HandleTransferSessionOutput(outEvent);
+        }
+        mTransfer.GetNextAction(outEvent);
+    }
+
+    mProcessingOutputEvents = false;
+
+    // If the mDestroySelf is set in NotifyEventHandled, we need to call DestroySelf() after processing all pending output events.
+    if (mDestroySelf)
+    {
+        DestroySelf();
+    }
+}
+
+CHIP_ERROR AsyncTransferFacilitator::SendMessage(const TransferSession::MessageTypeData msgTypeData,
+                                                 System::PacketBufferHandle & msgBuf)
+{
+    VerifyOrReturnError(mExchange, CHIP_ERROR_INCORRECT_STATE);
+
+    Messaging::SendFlags sendFlags;
+
+    // All messages that are sent expect a response, except for a StatusReport which would indicate an error and
+    // the end of the transfer.
+    if (!msgTypeData.HasMessageType(Protocols::SecureChannel::MsgType::StatusReport))
+    {
+        sendFlags.Set(Messaging::SendMessageFlags::kExpectResponse);
```

Comment:
Followup: instead of having logic in the caller that has to match this logic in terms of which types of messages expect responses, we should have either sendFlags or just the boolean of whether we set kExpectResponse be an outparam of this method.  But please don't make that change in this PR; do it separately.

Reply:
#36526. combined in the follow up issue